// browseth/dist/rpc/default.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var base_1 = require("./base");
var Default = (function (_super) {
    __extends(Default, _super);
    function Default(transport, endpoint, timeout, headers) {
        if (endpoint === void 0) { endpoint = 'http://localhost:8545'; }
        if (timeout === void 0) { timeout = 0; }
        var _this = _super.call(this) || this;
        _this.transport = transport;
        _this.endpoint = endpoint;
        _this.timeout = timeout;
        _this.headers = headers;
        return _this;
    }
    Default.prototype.handle = function (payload, cb) {
        this.transport.handle({
            url: this.endpoint,
            msg: JSON.stringify(payload),
            headers: __assign({}, this.headers, { 'Content-Type': 'application/json' }),
            timeout: this.timeout,
        }, function (err, response) {
            if (err) {
                cb(err);
                return;
            }
            try {
                cb(undefined, JSON.parse(response.msg));
            }
            catch (e) {
                cb(new Error("response is malformed: '" + response.msg + "'"));
            }
        });
    };
    return Default;
}(base_1.Rpc));
exports.Default = Default;
